### PR TITLE
[Owners] Adding system tests for Import and Export configuration buffer APIs

### PR DESCRIPTION
**Pull request by kruthi-NI**
### What does this Pull Request accomplish?

In the previous PRs #2 and #7 , I had [used reinterpret_cast in service_helpers.mako](https://github.com/doshirohan/grpc-device/blob/features/ni-dcpower/source/codegen/templates/service_helpers.mako#L190) file for [the input buffer of ViInt32[] and ViAddr[]] respectively. Also to handle the output buffer of ViAddr[] in `ExportAttributeConfigurationBuffer` API, I had modified mako file at [another location](https://github.com/doshirohan/grpc-device/blob/features/ni-dcpower/source/codegen/templates/service_helpers.mako#L218) by using reinterpret_cast.

Note that, in those 2 previous PRs, meta data was not added manually. It was just copy-pasted from base meta data. So what we want to test is not the meta data but we want to test the changes made to mako files at 2 locations (using reinterpret_cast).
ImportAttributeConfigurationBuffer and ExportAttributeConfigurationBuffer APIs cover both those changes made at 2 locations. Hence we decided to add system tests for `ImportAttributeConfigurationBuffer` and `ExportAttributeConfigurationBuffer` APIs in this PR.

### Why should this Pull Request be merged?

These tests helps to test the changes made to mako files for using reinterpret_cast.

### What testing has been done?

Manually ran system tests.

### CODE REVIEW

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 1]
```diff
@@ -231,7 +288,31 @@ TEST_F(NiDCPowerDriverApiTest, NiDCPowerSetViInt32Attribute_SendRequest_GetViInt
   EXPECT_EQ(expected_value, get_attribute_value);
 }
 
+TEST_F(NiDCPowerDriverApiTest, VoltageLevelConfiguredAndExportedToBuffer_ResetAndImportConfigurationFromBuffer_ConfigurationIsImportedSuccessfully)
```
Comment (kruthi-NI) [inline source/tests/system/nidcpower_driver_api_tests.cpp:407]: Note: Test is created this way:
1. First I have made some configurations to configure voltage level to 3.0
2. Then I have exported those configurations to a buffer
3. Then I have reset the configuration which will remove all the configurations made in step 1.
4. Then I have imported the configuration from the buffer and verified that configuration was imported successfully by checking that voltage level is set to 3.0 correctly

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 2]
```diff
@@ -324,6 +382,32 @@ TEST_F(NiDCPowerDriverApiTest, NiDCPowerSetViStringAttribute_SendRequest_GetViSt
...
+  {
+    auto configuration = export_buffer_response.configuration(i);
+    request.add_configuration(configuration);
+  }
+  dcpower::ImportAttributeConfigurationBufferResponse response;
+  ::grpc::Status status = GetStub()->ImportAttributeConfigurationBuffer(&context, request, &response);
+
+  EXPECT_TRUE(status.ok());
+  expect_api_success(response.status());
+  double actual_voltage_level = get_real64_attribute(channel_name, dcpower::NiDCPowerAttributes::NIDCPOWER_ATTRIBUTE_VOLTAGE_LEVEL);
```
Comment (ni-pkhare) [inline source/tests/system/nidcpower_driver_api_tests.cpp]: @kruthi-NI, using double will give expected result but for the sake of similarity can we use ViReal64 instead?

**Review by ganeshnie — APPROVED** [review]

**Review by ni-pkhare — APPROVED** [review]

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 2 above)
Comment (kruthi-NI) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: Done

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 3]
```diff
@@ -349,6 +407,32 @@ TEST_F(NiDCPowerDriverApiTest, NiDCPowerSetViInt64Attribute_SendRequest_GetViInt
...
+  {
+    auto configuration = export_buffer_response.configuration(i);
+    request.add_configuration(configuration);
+  }
+  dcpower::ImportAttributeConfigurationBufferResponse response;
+  ::grpc::Status status = GetStub()->ImportAttributeConfigurationBuffer(&context, request, &response);
+  EXPECT_TRUE(status.ok());
+  expect_api_success(response.status());
+
+  ViReal64 actual_voltage_level = get_real64_attribute(channel_name, dcpower::NiDCPowerAttributes::NIDCPOWER_ATTRIBUTE_VOLTAGE_LEVEL);
```
Comment (reckenro) [inline source/tests/system/nidcpower_driver_api_tests.cpp]: Does it make sense to also verify the output function was set by getting the `NIDCPOWER_ATTRIBUTE_OUTPUT_FUNCTION` attribute?

**Review by reckenro — APPROVED** [review]
Looks good. Just added a comment to add an extra assert for the output function attribute.

File: source/tests/system/nidcpower_driver_api_tests.cpp [hunk 4]
```diff
@@ -349,6 +407,32 @@ TEST_F(NiDCPowerDriverApiTest, NiDCPowerSetViInt64Attribute_SendRequest_GetViInt
...
+  ViReal64 expected_voltage_level = 3.0;
+  configure_output_function(channel_name, dcpower::OutputFunction::OUTPUT_FUNCTION_NIDCPOWER_VAL_DC_VOLTAGE);
+  configure_voltage_level(channel_name, expected_voltage_level);
+  auto export_buffer_response = export_attribute_configuration_buffer();
+
+  reset_with_channels(channel_name);
+  ::grpc::ClientContext context;
+  dcpower::ImportAttributeConfigurationBufferRequest request;
+  request.mutable_vi()->set_id(GetSessionId());
+  for (int i = 0; i < export_buffer_response.configuration_size(); i++)
```
Comment (jonlugoatx) [inline source/tests/system/nidcpower_driver_api_tests.cpp]: Use iterators to do the copy:
```cpp
auto exported_configuration = export_buffer_response.configuration();
import_request.add(exported_configuration.begin(), exported_configuration.end());
```
Then consider factoring out the copy step into a helper that takes in the export response & import request as parameters.

**Review by jonlugoatx — CHANGES_REQUESTED** [review]
The change looks good with a little bit of feedback. Marking as "Request changes" until the PR for the `users/pkhare/dcpower_setattribute_tests` branch goes in and you can target this PR at the feature branch.

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 3 above)
Comment (kruthi-NI) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: Done adding that assert

File: source/tests/system/nidcpower_driver_api_tests.cpp (hunk 4 above)
Comment (kruthi-NI) [reply source/tests/system/nidcpower_driver_api_tests.cpp]: @jonlugoatx 
I have tried to do the suggested change with minor modifications. Can you please have a look at the updated changes?

**Review by jonlugoatx — APPROVED** [review]

**Pull request merged**